### PR TITLE
[FEATURE] Joined layers are editable

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

In the context of the QEP qgis/QGIS-Enhancement-Proposals#27 in which the goal is to provide an auxiliary storage, a joined layer may be used to store auxiliary information. Thus, some new mechanisms are necessary to keep in sync the target layer and its joined layer(s).

This PR adds new options to configure a joined layer:
- *Editable* : a join layer may be updated from the target layer
- *Upsert on edit* : if a new feature is added in the target layer, a new feature is also added in the joined layer if necessary (without geometry?). Otherwise, if a joined feature already exists with the corresponding id, then the joined feature is just updated.
- *Delete cascade* : if a feature is deleted in the target layer, then the joined feature is deleted

joineditable

Some tests have been added to check the current behavior.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/qgsvectorlayer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayer.cpp]: What about if the join layer isn't editable? Should it be automatically made editable when the parent layer is set to editable if an editable join is present?

File: src/core/qgsvectorlayerjoininfo.h
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoininfo.h]: This name is a bit awkward. What about willUpsertOnEdit?

File: src/core/qgsvectorlayerjoininfo.h
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoininfo.h]: hasCascadedDelete? willDeleteCascaded?

File: src/core/qgsvectorlayerjoininfo.h
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoininfo.h]: Can you initialize these members here? (c++11 style)

**Review by nyalldawson — COMMENTED** [review]
This'll be fantastic to have :+1:

File: src/core/qgsvectorlayer.cpp
Comment (pblottiere) [inline src/core/qgsvectorlayer.cpp]: I've asked myself the same question. But I think it would be too permissive/dangerous to automatically set the joined layer as editable. However, I don't have a strong opinion here.

@haubourg What do you think about it?

Comment (haubourg) [inline src/core/qgsvectorlayer.cpp]: By myself, as a former data admin, I'm in favor of letting the user toggle editing by itself. That puts a fence in front of unwanted data modifications.

Comment (nyalldawson) [inline src/core/qgsvectorlayer.cpp]: That's fair enough - but in this case we need to give helpful feedback explaining why the edit failed. Silently failing isn't going to do us any favours with our users ;)

I'm also curious how this approach would work with the use case of joined columns for label data defined x/y (I can't find the relevant discussion now - but somewhere I remember a lengthy discussion about joins and automatically allowing movement of labels... any ideas where this is?).  Wouldn't this use case require automatic switching on of edit mode for the joined table containing x/y columns?

**Comment by nyalldawson** [comment]
Oops- forgot my other thoughts on this:

- "upsert" and "delete cascade" are VERY unfriendly terms for non-db admins to understand. I realise these are the correct names for these features, but is there a nicer term we can use here? If not, can we put some tooltips on all the widgets mentioning upsert/delete cascaded to explain them for novice users?

- I'm not convinced adding extra columns to this list will be scalable in the long term. I think we're going to quickly hit the issue we've got with the fields dialog. Instead of separate columns with ticks for each feature, what about a single column with a comma seperated list of enabled features? E.g. "Support dynamic forms, editable joins, cached in memory". It could be split over multiple lines too - horizontal space here is more precious then vertical space.

File: src/core/qgsvectorlayer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayer.cpp]: Ah - I'm blind. The discussion was in https://github.com/qgis/QGIS-Enhancement-Proposals/issues/27

Comment (haubourg) [inline src/core/qgsvectorlayer.cpp]: :smile: this PR is part of the that QEP27 yes !

**Comment by haubourg** [comment]
* "upsert": I would let that, because it's the right short option name and add a tooltip with "Insert, on conflict, update the row with the matching Id"

* UI design, good point, it's overcrowded now, we need a better layout. May I ask @nirvn an opinion on that Join Properties GUI (see the first screencast in the PR description) ? I know you're good at UI !

File: src/core/qgsvectorlayer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayer.cpp]: Exciting stuff... I can't wait for the benefits of this!

**Comment by nyalldawson** [comment]
> "Insert, on conflict, update the row with the matching Id"

What about "Automatically adds a matching row to the joined table, but if one already exists then update that matching row instead"?

**Comment by haubourg** [comment]
@nyalldawson that works too. I've had issues with long tooltips being cut on some OS or when raising the font size, so I was trying to make it shorter, but your version is a lot more user friendly. 
@pblottiere can you use that tooltip ?

File: src/core/qgsvectorlayer.cpp
Comment (pblottiere) [inline src/core/qgsvectorlayer.cpp]: > but in this case we need to give helpful feedback explaining why the edit failed. Silently failing isn't going to do us any favours with our users ;)

Indeed. I added an icon on the right of editor widgets in attribute form. Below an example where a feature is being created in the target layer with :
- **p1** not an editable join
- **p2** an editable join and *Upsert on edit* is ativated but the layer is not currently editable
- **p4** an editable join but *Upsert on edit* is not activated

@nyalldawson @haubourg Let me know of it suits you

Comment (pblottiere) [inline src/core/qgsvectorlayer.cpp]: > I'm also curious how this approach would work with the use case of joined columns for label data defined x/y/ Wouldn't this use case require automatic switching on of edit mode for the joined table containing x/y columns?

Actually, as mentionned in the QEP : 

> Auxiliary data may be updated by the user without updating source data, or for read only source data.

So even if it's true we'll have to automatically switch the edit mode of the joined table, it should not be linked to the edit mode of the target layer (which may be read only).

**Comment by pblottiere** [comment]
I added some tooltips:
- **Editable** : *This option allows values of the joined layers to be editable if they're themselves editable*
- **Upsert on edit** : *Automatically adds a matching row to the joined table, but if one already exists then update that matching row instead*
- **Delete cascade** : *Automatically delete the corresponding feature of the linked layer if one exists*

**Comment by pblottiere** [comment]
> I'm not convinced adding extra columns to this list will be scalable in the long term. I think we're going to quickly hit the issue we've got with the fields dialog. Instead of separate columns with ticks for each feature, what about a single column with a comma seperated list of enabled features? E.g. "Support dynamic forms, editable joins, cached in memory". It could be split over multiple lines too - horizontal space here is more precious then vertical space.

@nyalldawson What do you think about something like this (there's a lack in setting's item but it's just to confirm the idea):

**Comment by pblottiere** [comment]
@nyalldawson Can you give me your feedback on https://github.com/qgis/QGIS/pull/4913#issuecomment-318095396 and https://github.com/qgis/QGIS/pull/4913#discussion_r129507964 if you have a moment please?

Thank you :)

**Comment by m-kuhn** [comment]
@pblottiere is anything preventing you from merging this? doubts about the code or missing commit rights?

**Comment by pblottiere** [comment]
@m-kuhn I'm just finishing my work on auxiliary storage which is on top of this branch. Thus, if there's a lack in join editable stuff, I could update this PR. But it will be completed soon.

> missing commit rights?

I'm still waiting for the last opinion (Richard I think) before responding to the mailing list. So yes, still don't have commit rights ;)

**Comment by luipir** [comment]
great feature, tnx

**Comment by pblottiere** [comment]
I think this PR is ready to be merged. 

@nyalldawson @m-kuhn, can you do a review if you have some time?

Thank you!

File: src/core/qgsvectorlayer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayer.cpp]: I'd prefer a switch here (with no default clause)

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoinbuffer.cpp]: If you make QgsVectorLayerJoinBuffer a QgsFeatureSink (it already implements the required method) then you can do this method here (it's handled in QgsFestureSink)

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoinbuffer.cpp]: A possible performance improvement here:

You could set the no geometry flag on the request, and also request no attributes. That will speed up this check a lot. Then if the feature is found, use changeGeometry/changeAttributeValue instead of changeFeature (which also fires off a check for feature existence itself...)

Bonus points: add a `bool featureExists( QgsFeatureId id)` method in QgsVectorLayer to remove the duplicate code.

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (nyalldawson) [reply src/core/qgsvectorlayerjoinbuffer.cpp]: Do=drop

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (nyalldawson) [reply src/core/qgsvectorlayerjoinbuffer.cpp]: More thoughts: to do a featureExists(...) properly, it should be at the provider level, with a method in QgsVectorLayer which first checks the added/deleted buffers. Providers could bypass a lot of unnecessary feature parsing code if all that's needed is a check that a feature id is present. I'd only override the base class featureiterator approach for the OGR provider though, since that's what will benefit this use case.

Comment (pblottiere) [reply src/core/qgsvectorlayerjoinbuffer.cpp]: I set the no geometry flag and replace the *updateFeature* call by *changeAttributeValue* as you suggested. 

However, the aim of the feature request is to get a feature based on a column value (the join field name) thanks to an expression filter and not a feature id. So I don't understand how a *featureExists( bool fid )* method could be useful here. Can you enlighten me?

File: src/core/qgsvectorlayer.cpp
Comment (pblottiere) [reply src/core/qgsvectorlayer.cpp]: Done.

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (pblottiere) [reply src/core/qgsvectorlayerjoinbuffer.cpp]: Done

**Comment by pblottiere** [comment]
@nyalldawson Thank you for the review! Let me know what you think.

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (nyalldawson) [reply src/core/qgsvectorlayerjoinbuffer.cpp]: Ah, you're right. I read the code wrong.

File: src/core/qgsvectorlayerjoinbuffer.cpp
Comment (nyalldawson) [inline src/core/qgsvectorlayerjoinbuffer.cpp]: you should also setSubsetOfAttributes( QgsAttributeList() ) here - all the iterators have logic and tests to ensure that using a filter expression will always add the required attributes for the expression to be fetched.

**Review by nyalldawson — APPROVED** [review]

**Comment by pblottiere** [comment]
@nyalldawson I took into account your last comments. As you approved this PR, I'll merge it tomorrow if you have no more suggestions in the meantime.

Thank you!

**Comment by nyalldawson** [comment]
Looks great to me - merge away!

**Comment by m-kuhn** [comment]
Congratulations and welcome on board @pblottiere :tada:

**Comment by pblottiere** [comment]
Thank you @m-kuhn! I'll do my best to live up to the expectations!

**Comment by nyalldawson** [comment]
@pblottiere I suspect this PR (or one of the related ones) has caused a crasher in 3.0/3.1. Steps to reproduce are simple:

- open an attribute table
- make the layer editable
- try to delete the last field in the layer

I get a crash (on master) at qgsattributeform.cpp:2015

**Comment by pblottiere** [comment]
@nyalldawson Indeed. Here's the bugfix: https://github.com/qgis/QGIS/pull/6892

**Pull request merged**
